packages/block-editor/src/components/block-navigation/block-contents.js

```diff
@@ -0,0 +1,185 @@
+	getBlockType,
+} from '@wordpress/blocks';
+import { Button, Fill, Slot, VisuallyHidden } from '@wordpress/components';
+import { useInstanceId } from '@wordpress/compose';
+import {
+	Children,
+	cloneElement,
+	forwardRef,
+	useContext,
+} from '@wordpress/element';
+import { __, sprintf } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import BlockIcon from '../block-icon';
+import { BlockListBlockContext } from '../block-list/block';
+import { useBlockNavigationContext } from './context';
+
+export const BlockNavigationBlockContentWrapper = forwardRef( function(
+	{
+		as: WrapperComponent,
+		className,
+		block,
+		isSelected,
+		onClick,
+		position,
+		siblingCount,
+		level,
+		children,
+		...props
+	},
+	ref
+) {
+	const { name, attributes } = block;
+	const instanceId = useInstanceId( BlockNavigationBlockContentWrapper );
+	const descriptionId = `block-navigation-block-select-button_${ instanceId }`;
+	const blockType = getBlockType( name );
+	const blockDisplayName = getBlockLabel( blockType, attributes );
+	const blockPositionDescription = sprintf(
+		/* translators: 1: The numerical position of the block. 2: The total number of blocks. 3. The level of nesting for the block. */
+		__( 'Block %1$d of %2$d, Level %3$d' ),
+		position,
+		siblingCount,
+		level
+	);
+
+	return (
+		<>
+			<WrapperComponent
+				className={ classnames(
+					'block-editor-block-navigation-block-content-wrapper',
+					className
+				) }
+				onClick={ onClick }
+				aria-describedby={ descriptionId }
+				ref={ ref }
+				{ ...props }
+			>
+				<BlockIcon icon={ blockType.icon } showColors />
+				{ children ? children : blockDisplayName }
+				{ isSelected && (
+					<VisuallyHidden>
+						{ __( '(selected block)' ) }
+					</VisuallyHidden>
+				) }
+			</WrapperComponent>
+			<div
+				className="block-editor-block-navigation-block-content-wrapper__description"
+				id={ descriptionId }
+			>
+				{ blockPositionDescription }
+			</div>
+		</>
+	);
+} );
+
+const BlockNavigationBlockSelectButton = forwardRef( ( props, ref ) => (
+	<BlockNavigationBlockContentWrapper
+		ref={ ref }
+		as={ Button }
+		className="block-editor-block-navigation-block-select-button"
+		{ ...props }
+	/>
+) );
+
+const getSlotName = ( clientId ) => `BlockNavigationBlock-${ clientId }`;
+
+const BlockNavigationBlockSlot = forwardRef(
+	(
+		{ block, isSelected, onClick, position, siblingCount, level, ...props },
+		ref
+	) => {
+		const { clientId } = block;
+
+		return (
+			<Slot name={ getSlotName( clientId ) }>
+				{ ( fills ) => {
+					if ( ! fills.length ) {
+						return (
+							<BlockNavigationBlockSelectButton
+								ref={ ref }
+								block={ block }
+								onClick={ onClick }
+								isSelected={ isSelected }
+								position={ position }
+								siblingCount={ siblingCount }
+								level={ level }
+								{ ...props }
+							/>
+						);
+					}
+
+					return (
+						<BlockNavigationBlockContentWrapper as="div">
```

Comment:
@adamziel This main file now contains a lot of the Wrapper and Slot/Fill logic. I ended up renaming a few things.

The main logic I changed is that the wrapper is now rendered around the slot, while in your PR the wrapper is rendered inside the fill:
https://github.com/WordPress/gutenberg/pull/22210/files#diff-2e071f8429cd2c500b830947ffde4239R226-R231

Pretty open to handling that differently, I think some aspects of this may need to be changed to handle adding the menu described in https://github.com/WordPress/gutenberg/issues/22089.

Reply:
@talldan I went for the "wrapper inside od the fill" approach to make something like this possible:

```js
<BlockNavigationBlockFill>
	<BlockNavigationBlockSelectButton
		ref={ ref }
		block={ block }
		onClick={ onClick }
		isSelected={ isSelected }
		position={ position }
		siblingCount={ siblingCount }
		level={ level }
		{ ...props }
	/>
	<EllipsisMenu {/* some props */} />
</BlockNavigationBlockFill>
```

With "wrapper outside of the fill" approach, these slots are only good for something like replacing the label with RichText. You cannot update/replace the button AND add something before/after a button.

Reply:
Hmm, this is made difficult because this PR introduces the movers which are rendered after the fill, and then the designs in #22089 shown the dropdown menu after the movers.

I think most blocks will end up having the dropdown menu, so it probably makes sense to add that as an internal part of `BlockNavigation`, and then if we need to add custom items for particular blocks we can try a slot for the menu.